enhancements/cvo/cluster-profiles.md

```diff
@@ -0,0 +1,103 @@
+---
+title: Cluster Profiles
+authors:
+  - "@csrwng"
+reviewers:
+  - "@abhinavdahiya"
+  - "@crawford"
+  - "@sdodson"
+  - "@derekwaynecarr"
+  - "@smarterclayton"
+approvers:
+  - "@derekwaynecarr"
+  - "@smarterclayton"
+creation-date: 2020-02-04
+last-updated: 2019-02-04
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# cluster-profiles
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+Cluster profiles are a way to support different deployment models for OpenShift clusters. 
+A profile is an identifier that the Cluster Version Operator uses to determine
+which manifests to apply. Operators can be excluded completely or can have different
+manifests for each supported profile.
+
+## Motivation
+
+In order to support different deployment models in which not all operators rendered by
+the CVO by default are needed, a mechanism is needed by which we tell the CVO which 
+manifests to include or exclude. One such deployment is the hosted control plane case
+in which the control plane components and corresponding operators should not be rendered
+by the CVO.
+
+
+### Goals
+
+- Equip the CVO to include/exclude manifests based on a matching identifier (profile)
+
+### Non-Goals
+
+- Define which manifests should be excluded/included for each use case
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+As a user, I can create a cluster in which manifests for control plane operators are
+not applied by the CVO.
+
+#### Story 2
+As a user, I can create a cluster in which node selectors for certain operators target
+worker nodes instead of master nodes.
+
```

Comment:
As per @enxebre perhaps something like

#### Story 3
As a user, I can create a cluster for the `baremetal` platform where the `baremetal-operator` manifests are applied by the CVO.

Reply:
For more context on that, see #212

Reply:
Hrm - this is definitely scope expansion.  In general, profiles don't seem appropriate for conditional parameterization of the payload based on global configuration (which platform type is), because a second level operator can handle that.  No matter what, there needs to be a second level operator in the payload for handling deployment of the resources necessary for machine api that are part of the platform.

I'll look at 212, but I would probably say that machine-api-operator is responsible for deploying a sub operator to manage bare metal if one is truly necessary.

Reply:
The general problem with profiles is that they expand the scope of the templating the payload provides.  The current profile proposes exactly 2 variants decided at install time, and they impact how the core operator works as well as how SLOs are deployed or included.  If we expanded this to include operators that are determined by infrastructure, then we're potentially introducing a new variable (not just a new profile), since we very well may want to deploy bare metal operator in a hypershift mode.

I don't want to add functionality to profiles until we have a concrete need, and given hyper shift is not ready and is not going to be near term ready, I'd prefer to focus solving BMO with extensions of our existing approach, not new approaches.

Reply:
agree w/ @smarterclayton .  hypershift style deployments which is the primary profile being proposed here would potentially want bare metal machine management in the future.

Reply:
On the "hypershift would potentially want bare metal machine management in the future" point - I acknowledged this in the point below about needing to allow multiple cluster profiles to be combined if we want to support this use case.

But ok, feedback taken - #212 won't pursue the cluster profile path, instead it will propose a new SLO for bare metal.

Reply:
(can be resolved)